File: IrisClassificationStudy/C++Version/IrisClassificationStudy/code/Perceptron.cpp
//
//  Perceptron.cpp
//  IrisClassificationStudy
//
//  Created by ivastival on 22/11/21.
//

#include "Perceptron.hpp"

Perceptron::Perceptron(){
    this->seed = 45;
    this->eta  = 0.01;
    this->iter_num = 50;
}

Perceptron::Perceptron(int seed, float eta, int iter_num){
    this->seed = seed;
    this->eta  = eta;
    this->iter_num = iter_num;
}

Perceptron::~Perceptron(){
    delete this->W;
}

void Perceptron::fit(const vector< vector<float> >& X, const vector<int>& Y){
    
    // SET THE WEIGHTS SIZE, IT'S THE TOTAL COLUMNS (CHARACTERISTICS) + THE BIAS WEIGHT
    unsigned long var_number = X[0].size()+1;
    this->W = new vector<float>(var_number,0);
    
    srand(1);
    
    // POPULATE WITH RANDOM VALUES
    std::generate(this->W->begin(), this->W->end(), [](){
        return ((rand() % 10)/ 5.0) + 0.01;
    });
    
    float update = 0.0;
    int error;
    
    for(int i=0; i < this->iter_num; ++i){
        vector<vector<float>>::const_iterator x_it = X.begin();
        vector<int>::const_iterator y_it = Y.begin();
        error = 0.0;
        
        for(; x_it != X.end(); ++x_it, ++y_it){
            update = this->eta * (*y_it - this->predict(*x_it));
            
            if(update != 0) error += 1;
            
            this->updateWeights(update, *x_it);
        }
        this->errors.push_back(error);
    }
//    copy(this->W->begin(), this->W->end(), std::ostream_iterator< float >(cout, ", "));
}

// PREDICT RETURN A SIMPLE PREDTICTION RESULT IF THE net_input() IS BIGGER OR EQUAL TO 0 RETURN 1 ELSE RETURN -1
int Perceptron::predict(const vector<float> &X){
    if(this->net_input(X) >= 0.0) return 1; else return -1;
}

// LINEAR FUNCTION BETWEEN THE X DATA CHARACTERISTICS AND RANDOM GENERATED WEIGHTS
float Perceptron::net_input(const vector<float> &X){

    vector<float> w_temp(this->W->begin() + 1, this->W->end());
    float result = UtilsMath::dot(X, w_temp) + (*this->W)[0];
    return result;
}

// FUNCTION TO UPDATE THE WEIGHTS BASED ON THE ERROR OF PREDICT
void Perceptron::updateWeights(float update, const vector<float> &X){
    // UPDATE THE BIAS WEIGHT
    (*this->W)[0] += update;
    
    vector<float>::iterator w_it = this->W->begin();
    vector<float>::const_iterator x_it = X.begin();

    // UPDATE EACH CHARACTERISTICS WEIGHTS
    vector<float> w_temp(this->W->begin() + 1, this->W->end());
    for(;w_it != this->W->end(); ++w_it, ++x_it){
        *w_it += update * *x_it;
    }
}


void Perceptron::printErrors(){
    copy(this->errors.begin(), this->errors.end(), std::ostream_iterator<int>(cout, " ,"));
}

File: IrisClassificationStudy/PythonVersion/code/Perceptron.py
import pandas as pd
import numpy as np



class Perceptron:
    def __init__(self, seed=45, eta=0.01, iter_num=10):

        self.seed = seed
        self.eta  = eta
        self.iter_num = iter_num

    def fit(self, X, Y):

        rgen = np.random.RandomState(self.seed)
        self.w_ = rgen.normal(loc=0.0, scale=0.01, size=1 + X.shape[1])

        self.errors = []

        for _ in range(self.iter_num):
            error = 0
            for xi, target in zip(X, Y):
                update = self.eta * (target - self.predict(xi))
                print(self.predict(xi))
                print(xi)

                self.w_[1:] += update * xi
                self.w_[0]  += update

                error += int(update != 0.0)
            self.errors.append(error)
        return self

    def net_input(self, X):
        return np.dot(X, self.w_[1:]) + self.w_[0]

    def predict(self, X):
        return np.where(self.net_input(X) >= 0.0, 1 , -1)
